esmvalcore/_task.py

```diff
@@ -577,11 +582,61 @@ def _collect_provenance(self):
             product = TrackedFile(filename, attributes, ancestors)
             product.initialize_provenance(self.activity)
             product.save_provenance()
+            self._write_citation_file(product)
             self.products.add(product)
         logger.debug("Collecting provenance of task %s took %.1f seconds",
                      self.name,
                      time.time() - start)
 
+    @staticmethod
+    def _write_citation_file(product):
+        """
+            Write citation information provided by the recorded provenance.
+            Recipe and cmip6 data references are saved into one bibtex file.
+            cmip6 data references are provided by CMIP6 data citation service.
+            each cmip6 data reference has a json link. In the case of internet
+            connection, cmip6 data references are saved into a bibtex file.
+            Otherwise, cmip6 data reference links are saved into a text file.
+        """
+        # collect info from provenance
+        file_name = os.path.splitext(product.filename)[0]
+        citation = {
+            'references': [],
+            'info_urls': [],
+            'tag': [],
+            'entry': '',
+            'url': '',
+            'file': [
+                file_name + '_data_citation_url.txt',
+                file_name + '_citation.bibtex',
+            ]}
+        for item in product.provenance.records:
+            for key, value in item.attributes:
+                if key.namespace.prefix == 'attribute':
+                    if key.localpart in {'reference', 'references'}:
+                        citation['references'].append(value)
+                    elif key.localpart == 'further_info_url':
+                        citation['info_urls'].append(value)
+
+        # collect CMIP6 citation, if any
+        if citation['info_urls']:
+            citation['entry'], citation['url'] = _collect_cmip_citation(
+                citation['info_urls'])
+        if citation['url']:
+            with open(citation['file'][0], 'w') as file:
+                file.write(citation['url'])
+
+        # collect recipe citation, if any
+        if citation['references']:
+            citation['tag'] = _replace_entry(citation['references'])
```

Comment:
Instead of doing this reverse lookup, it would be easier to just replace the `references` tags with the values after using them to find the associated bibtex file. The content of the bibtex file should then be used to construct a meaningful short description.